File: Elevenstudents/Flashcards/tools/apkg_to_json.py
import json
import os
import sqlite3
import zipfile
from pathlib import Path

# Usage: python apkg_to_json.py
# Converts .apkg files in Flashcards/ANKI files/ to JSON in Flashcards/data/

ROOT = Path(__file__).resolve().parents[1]
APKG_DIR = ROOT / 'ANKI files'
OUT_DIR = ROOT / 'data'
OUT_DIR.mkdir(parents=True, exist_ok=True)

DECKS = {
    'biology': 'IGCSE__Biology_v2.apkg',
    'chemistry': 'IGCSE__Chemistry_v3.apkg',
    'physics': 'IGCSE__Physics_v1.apkg',
    'physics-eqns': 'IGCSE__Physics-eqns_v1.apkg',
}


def extract_collection(apkg_path: Path, work_dir: Path) -> Path:
    """Extract collection.anki2/collection.anki21 from the .apkg into work_dir and return path."""
    with zipfile.ZipFile(apkg_path, 'r') as z:
        name = None
        if 'collection.anki2' in z.namelist():
            name = 'collection.anki2'
        elif 'collection.anki21' in z.namelist():
            name = 'collection.anki21'
        else:
            raise FileNotFoundError('collection.anki2/collection.anki21 not found in apkg')
        out_path = work_dir / name
        with z.open(name) as src, open(out_path, 'wb') as dst:
            dst.write(src.read())
        return out_path


def read_cards_from_collection(collection_path: Path):
    conn = sqlite3.connect(str(collection_path))
    try:
        cur = conn.cursor()
        cur.execute(
            '''SELECT cards.id, notes.flds FROM cards JOIN notes ON cards.nid = notes.id'''
        )
        cards = []
        for card_id, flds in cur.fetchall():
            # Anki field separator is \x1f
            parts = flds.split('\x1f')
            front = parts[0] if parts else ''
            back = parts[1] if len(parts) > 1 else ''
            cards.append({
                'id': card_id,
                'front': front,
                'back': back,
            })
        return cards
    finally:
        conn.close()


def convert_all():
    tmp_dir = OUT_DIR / '_tmp'
    tmp_dir.mkdir(parents=True, exist_ok=True)
    for deck_id, filename in DECKS.items():
        apkg_path = APKG_DIR / filename
        if not apkg_path.exists():
            print(f'[WARN] Missing apkg: {apkg_path}')
            continue
        try:
            coll_path = extract_collection(apkg_path, tmp_dir)
            cards = read_cards_from_collection(coll_path)
            out_file = OUT_DIR / f'{deck_id}.json'
            with open(out_file, 'w', encoding='utf-8') as f:
                json.dump(cards, f, ensure_ascii=False, indent=2)
            print(f'[OK] {deck_id}: {len(cards)} cards -> {out_file}')
        except Exception as e:
            print(f'[ERROR] {deck_id}: {e}')
    # Clean up temp files
    for p in tmp_dir.glob('collection.anki*'):
        try:
            os.remove(p)
        except Exception:
            pass


if __name__ == '__main__':
    convert_all()

File: Elevenstudents/Flashcards/tools/apkg_to_json.js
import AdmZip from 'adm-zip';
import initSqlJs from 'sql.js';
import { readFileSync, writeFileSync, mkdirSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const ROOT = join(__dirname, '..');
const APKG_DIR = join(ROOT, 'ANKI files');
const OUT_DIR = join(ROOT, 'data');

const DECKS = {
  'biology': 'IGCSE__Biology_v2.apkg',
  'chemistry': 'IGCSE__Chemistry_v3.apkg',
  'physics': 'IGCSE__Physics_v1.apkg',
  'physics-eqns': 'IGCSE__Physics-eqns_v1.apkg',
};

// Strip HTML tags
function stripHtml(html) {
  return html.replace(/<[^>]*>/g, '').trim();
}

async function convertApkg(deckId, filename) {
  const apkgPath = join(APKG_DIR, filename);
  console.log(`\nProcessing ${deckId}: ${filename}`);
  
  try {
    // Extract .apkg (it's a zip file)
    const zip = new AdmZip(apkgPath);
    let collectionEntry = zip.getEntry('collection.anki2') || zip.getEntry('collection.anki21');
    
    if (!collectionEntry) {
      console.error(`  ❌ No collection.anki2/anki21 found`);
      return;
    }
    
    const dbData = collectionEntry.getData();
    
    // Load SQLite database
    const SQL = await initSqlJs();
    const db = new SQL.Database(dbData);
    
    // Query cards
    const result = db.exec(`
      SELECT cards.id, notes.flds
      FROM cards
      JOIN notes ON cards.nid = notes.id
    `);
    
    if (!result.length) {
      console.error(`  ❌ No cards found`);
      return;
    }
    
    const cards = [];
    result[0].values.forEach(([cardId, flds]) => {
      const fields = flds.split('\x1f'); // Anki field separator
      if (fields.length >= 2) {
        cards.push({
          id: cardId,
          front: stripHtml(fields[0]),
          back: stripHtml(fields[1])
        });
      }
    });
    
    db.close();
    
    // Write JSON
    mkdirSync(OUT_DIR, { recursive: true });
    const outPath = join(OUT_DIR, `${deckId}.json`);
    writeFileSync(outPath, JSON.stringify(cards, null, 2));
    
    console.log(`  ✅ Exported ${cards.length} cards → ${deckId}.json`);
  } catch (error) {
    console.error(`  ❌ Error: ${error.message}`);
  }
}

async function main() {
  console.log('Converting Anki decks to JSON...\n');
  
  for (const [deckId, filename] of Object.entries(DECKS)) {
    await convertApkg(deckId, filename);
  }
  
  console.log('\n✅ Done!');
}

main();
